### ISSUE TITLE
`default_constructed_unit_structs` shouldn't apply to unit structs outside the current crate

### ISSUE THREAD

**Issue by davidbarsky** [rust-lang/rust-clippy#11315]
### Summary

Hi folks! This isn't really a bug report, but rather, feedback. The summary of the feedback is that `default_constructed_unit_structs` lint took a semver mistake I made in a crate—[`tracing-glog`](https://docs.rs/tracing-glog/)—and made it a much more punishing mistake when I was upgrading my employer's monorepo from `tracing-glog-0.2.3` to `tracing-glog-0.3.0`. Here's how:

The documentation for `tracing-glog` recommends that people set it as the following:

```rust
use tracing_glog::{Glog, GlogFields};

tracing_subscriber::fmt()
    .event_format(Glog::default())
    .fmt_fields(GlogFields::default())
    .init();
```

With the release of Rust 1.71, Clippy pushed people to rewrite that as:

```rust
use tracing_glog::{Glog, GlogFields};

tracing_subscriber::fmt()
    .event_format(Glog::default())
    .fmt_fields(GlogFields)
    .init();
```

However, [a PR](https://github.com/davidbarsky/tracing-glog/pull/15) added a private field at the roughly the same same time that 1.71 was released. This (unfortunate!) coincidence turned what was a mistake in the API—`GlogFields` being able to be constructed as a unit struct—it something that made the monorepo-wide upgrade harder and more complicated than it needed to be.

My feeling is that while this lint is pretty useful, it should be scoped to unit structs _within the current crate_, as there is no semver hazard there. Applying it to external crates can be too punishing from a semver hazard standpoint.

(Note that the commit hash might be not what you expect, as we carry a few patches to `rustc` that include things like "use `jemalloc` as the default allocator", but nothing that should change Clippy's behavior)

### Reproducer

[`GlogFields`](https://docs.rs/tracing-glog/0.2.2/tracing_glog/struct.GlogFields.html) in `tracing-glog-0.2.2`

### Version

```text
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
I don't really agree. You can annotate these unit structs with `#[non_exhaustive]` if there are no guarantees it'll remain one, which prevents you from constructing it at all (Also, TIL you can do that. Maybe clippy should suggest that, somehow :thinking:)

You can't call a function that takes `&self` like that, but you can call `Default::default` or `new` (basically, if it doesn't take `Self`).

**Comment by Centri3** [comment]
It's unfortunate though that it seems `non_exhaustive` on unit structs is only ever brought up once in the reference (and it's not even the struct in particular; it's an enum variant...). So, I really think that should be expanded on
